### PR TITLE
CosmWasmClient types

### PR DESCRIPTION
**Pull request by webmaster128**
Create REST API independent higher level types for all CosmWasmClient interfaces.

~Based on #129~

### CODE REVIEW

File: packages/sdk/src/cosmwasmclient.searchtx.spec.ts [hunk 1]
```diff
@@ -92,8 +92,8 @@ describe("CosmWasmClient.searchTx", () => {
       expect(result.length).toEqual(1);
       expect(result[0]).toEqual(
         jasmine.objectContaining({
-          height: postedSend.height.toString(),
-          txhash: postedSend.hash,
+          height: postedSend.height,
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.searchtx.spec.ts]: height as number, and rename. This definitely looks nicer

File: packages/sdk/src/cosmwasmclient.spec.ts [hunk 2]
```diff
@@ -83,10 +83,10 @@ describe("CosmWasmClient", () => {
       const client = new CosmWasmClient(wasmdEndpoint);
       expect(await client.getAccount(unused.address)).toEqual({
         address: unused.address,
-        account_number: 5,
+        accountNumber: 5,
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.spec.ts]: No more eslint:camel-case exceptions :+1:

File: packages/sdk/src/cosmwasmclient.ts [hunk 3]
```diff
@@ -92,6 +102,39 @@ export interface ContractDetails extends Contract {
...
+  readonly height: number;
+  readonly hash: string;
+  readonly rawLog: string;
+  readonly logs: readonly Log[];
+  readonly tx: CosmosSdkTx;
+  /** The gas limit as set by the user */
+  readonly gasWanted?: number;
+  /** The gas used by the execution */
+  readonly gasUsed?: number;
+  readonly timestamp: string;
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.ts]: Please doc what the string format is (I assume same as BlockHeader.time?) There is an issue when parsing in go where you must know if it is supposed to be second, millisecond, microsecond, or nanosecond precision....

File: packages/sdk/src/cosmwasmclient.ts [hunk 4]
```diff
@@ -92,6 +102,39 @@ export interface ContractDetails extends Contract {
   readonly initMsg: object;
 }
 
+/** A transaction that is indexed as part of the transaction history */
+export interface IndexedTx {
+  readonly height: number;
+  readonly hash: string;
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.ts]: This is upper-case hex, right?

File: packages/sdk/src/cosmwasmclient.ts [hunk 5]
```diff
@@ -2,14 +2,24 @@ import { Sha256 } from "@iov/crypto";
...
 export interface GetNonceResult {
   readonly accountNumber: number;
   readonly sequence: number;
 }
 
+export interface Account {
+  /** Bech32 account address */
+  readonly address: string;
+  readonly balance: ReadonlyArray<Coin>;
+  /** Bech32 encoded pubkey */
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.ts]: I find the pubkey object (algorithm, data) format easier to handle outside of this package. Unless we decide bech32 pubkey is the only public interface everywhere (like from the Pen), I would prefer to use the object version for consistency, and just use the bech32 internally.

Comment (ethanfrey) [reply packages/sdk/src/cosmwasmclient.ts]: eg, if I want to check the algorithm type... do I want to grab a few bytes and check for the amino-encoding in some file? Would anyone but us two even get that idea?

File: packages/sdk/src/cosmwasmclient.ts [hunk 6]
```diff
@@ -129,31 +172,47 @@ export class CosmWasmClient {
...
-      return this.restClient.blocks(height);
-    } else {
-      return this.restClient.blocksLatest();
-    }
+  public async getBlock(height?: number): Promise<Block> {
+    const response =
+      height !== undefined ? await this.restClient.blocks(height) : await this.restClient.blocksLatest();
+
+    return {
+      id: response.block_id.hash,
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.ts]: This is very much nicer

File: packages/sdk/src/cosmwasmclient.ts [hunk 7]
```diff
@@ -129,31 +172,47 @@ export class CosmWasmClient {
...
+  public async getAccount(address: string): Promise<Account | undefined> {
     const account = await this.restClient.authAccounts(address);
     const value = account.result.value;
-    return value.address === "" ? undefined : value;
+    return value.address === ""
+      ? undefined
+      : {
+          address: value.address,
+          balance: value.coins,
+          pubkey: value.public_key || undefined,
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.ts]: Nice cleanup. I would consider converting the public_key from bech32 into an object (as mentioned above). I am pretty sure the parse code exists

File: packages/sdk/src/cosmwasmclient.ts [hunk 8]
```diff
@@ -313,6 +369,15 @@ export class CosmWasmClient {
         `Found more results on the backend than we can process currently. Results: ${result.total_count}, supported: ${limit}`,
       );
     }
-    return result.txs;
+    return result.txs.map(
+      (restItem): IndexedTx => ({
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.ts]: Nice and much more standardized

File: packages/sdk/types/signingcosmwasmclient.d.ts [hunk 9]
```diff
@@ -28,6 +28,16 @@ export interface UploadReceipt {
   readonly compressedChecksum: string;
   /** The ID of the code asigned by the chain */
   readonly codeId: number;
+  readonly logs: readonly Log[];
+  /** Transaction hash (might be used as transaction ID). Guaranteed to be non-empty upper-case hex */
+  readonly transactionHash: string;
+}
+export interface InstantiateResult {
```
Comment (ethanfrey) [inline packages/sdk/types/signingcosmwasmclient.d.ts]: Nice to add more into here as well

File: packages/sdk/src/signingcosmwasmclient.ts [hunk 10]
```diff
@@ -188,7 +200,11 @@ export class SigningCosmWasmClient extends CosmWasmClient {
 
     const result = await this.postTx(signedTx);
     const contractAddressAttr = findAttribute(result.logs, "message", "contract_address");
-    return contractAddressAttr.value;
+    return {
```
Comment (ethanfrey) [inline packages/sdk/src/signingcosmwasmclient.ts]: Very cool

File: packages/sdk/src/signingcosmwasmclient.ts [hunk 11]
```diff
@@ -150,6 +160,8 @@ export class SigningCosmWasmClient extends CosmWasmClient {
       compressedSize: compressed.length,
       compressedChecksum: Encoding.toHex(new Sha256(compressed).digest()),
       codeId: Number.parseInt(codeIdAttr.value, 10),
+      logs: result.logs,
```
Comment (ethanfrey) [inline packages/sdk/src/signingcosmwasmclient.ts]: Looks like all results get normalized with logs and transactionHash. I like that

File: packages/sdk/src/cosmwasmclient.ts (hunk 5 above)
Comment (webmaster128) [reply packages/sdk/src/cosmwasmclient.ts]: Good point

File: packages/sdk/src/cosmwasmclient.ts (hunk 4 above)
Comment (webmaster128) [reply packages/sdk/src/cosmwasmclient.ts]: Jap, documented in 23f88564e5df7ec092b9caf76708cb26d1cd12bc

File: packages/sdk/src/cosmwasmclient.ts [hunk 12]
```diff
@@ -185,7 +187,7 @@ export class CosmWasmClient {
       : {
           address: value.address,
           balance: value.coins,
-          pubkey: value.public_key || undefined,
+          pubkey: value.public_key ? decodeBech32Pubkey(value.public_key) : undefined,
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.ts]: :+1:

**Pull request merged**